symbolic-debuginfo/src/dwarf.rs

```diff
@@ -1320,6 +1344,31 @@ impl<'s> Iterator for DwarfUnitFileIterator<'s> {
     }
 }
 
+fn resolve_byte_name<'s>(bcsymbolmap: Option<&BcSymbolMap<'s>>, s: &'s [u8]) -> &'s [u8] {
+    if let Some(bcsymbolmap) = bcsymbolmap {
+        std::str::from_utf8(s)
+            .map(|s| bcsymbolmap.resolve(s))
+            .ok()
+            .map(|s| s.as_bytes())
+            .unwrap_or(s)
+    } else {
+        s
+    }
+}
+
+fn resolve_cow_name<'s>(bcsymbolmap: Option<&BcSymbolMap<'s>>, s: Cow<'s, str>) -> Cow<'s, str> {
+    if let Some(bcsymbolmap) = bcsymbolmap {
+        match s {
+            Cow::Borrowed(s) => bcsymbolmap.resolve(s).into(),
+            // we make an unconditional copy here, since the borrow checker has no idea
+            // if the returned `str` is owned by the `bcsymbolmap`, or by `s`.
+            Cow::Owned(s) => bcsymbolmap.resolve(&s).to_owned().into(),
```

Comment:
Why do you convert this branch to owned, but not the other? You can always borrow from the symbolmap. IMHO, this entire function should just be (inlined):

```rust
bcsymbolmap.map_or(s, |m| m.resolve(&s).into())
```

Reply:
`resolve()` is bound to the lifetime of the argument, since it might return the argument unmodified. Thats what I mean by the borrow checker can’t tell the difference.

Reply:
Thanks, I missed that. Looking at [`BcSymbolMap`](https://docs.rs/symbolic/8.2.1/symbolic/debuginfo/macho/struct.BcSymbolMap.html#implementations), we might actually want to extend its API surface indeed. We have a `get(usize)` which returns an `Option`, but no `resolve` that does so. Maybe what we want is:

- `get(&self, usize) -> Option<&'d str>`
- `resolve(&self, &str) -> Option<&'d str>`
- `resolve_default(&self, &'d str) -> &'d str` (note that this is binding to the same `'d` as the data)

Not sure about the binary interface, though. Probably, it's fine to assume you only want to pass UTF-8 strings.